src/app.cpp

```diff
@@ -116,6 +116,37 @@ class CanFrame {
 
 };
 
+template<class ObjectHeaderGeneric>
+std::uint64_t calculate_ts_res(ObjectHeaderGeneric *oh)
+{
+	uint64_t ts_resol = 0;
+	switch (oh->objectFlags) {
+	case ObjectHeader::ObjectFlags::TimeTenMics:
+		ts_resol = 100000;
+	case ObjectHeader::ObjectFlags::TimeOneNans:
+		ts_resol = NANOS_PER_SEC;
+		break;
+	default:
+		fprintf(stderr, "ERROR: The timestamp format is unknown (not 10us nor ns)!\n");
+		break;
+	}
+	return ts_resol;
+}
+
+template<class ObjectHeaderGeneric>
+pcapng_exporter::frame_header generate_header(
+	ObjectHeaderGeneric *oh,
+	std::uint64_t date_offset_ns)
+{
+	pcapng_exporter::frame_header header = pcapng_exporter::frame_header();
+	header.channel_id = oh->channel;
+	header.timestamp_resolution = calculate_ts_res(oh);
+	uint64_t ts = (NANOS_PER_SEC / header.timestamp_resolution) * oh->objectTimeStamp + date_offset_ns;
+	header.timestamp.tv_sec = ts / NANOS_PER_SEC;
+	header.timestamp.tv_nsec = ts % NANOS_PER_SEC;
+	return header;
```

Comment:
Duplicate block with write_packet

Reply:
Not exactly. The header types differ (light_pcap_header and frame_header; lightpcap vs pcapng).
It is basically possible to unify this funcftionality using another template function, but this enhances the risk of run-time failures and I discourage the use of template functions unless absolutely necessary.